README.md
# contributors-from-git

> **Get an array of contributors from a git working tree.**  
> Requires git 1.7+. Forked from [`git-contributors`](https://github.com/davidlinse/git-contributors.js) to reduce its scope (no output formatters, command-line interface, sorting or filtering).

[![npm](https://img.shields.io/npm/v/contributors-from-git.svg)](https://www.npmjs.com/package/contributors-from-git)
[![Node version](https://img.shields.io/node/v/contributors-from-git.svg)](https://www.npmjs.com/package/contributors-from-git)
[![Build Status](https://travis-ci.org/vweevers/contributors-from-git.svg)](https://travis-ci.org/vweevers/contributors-from-git)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com) [![Greenkeeper badge](https://badges.greenkeeper.io/vweevers/contributors-from-git.svg)](https://greenkeeper.io/)

## Usage

```
npm install contributors-from-git
```

```js
var contributors = require('contributors-from-git')

contributors('.', function (err, result) {
  if (err) throw err
  console.log(result)
})
```

This yields an array of contributors:

```js
[
  { commits: 40, name: 'Maja',  email: 'maja@hive' },
  { commits: 10, name: 'Flip',  email: 'flip@meadow' },
  { commits: 80, name: 'Willi', email: 'willi@sunflower' }
]
```

## API

### `contributors(dir, callback)`

The `dir` argument must resolve to a git working tree.

## License

[MIT](LICENSE) © 2019 Vincent Weevers, © 2014 David Linse.


CHANGELOG.md
<a name="0.2.3"></a>
### 0.2.3 (2015-06-14)

* minor refactorings


<a name="0.2.2"></a>
### 0.2.2 (2015-06-08)

* maintenance release including minor fixes and updated dependencies


<a name="0.2.0"></a>
## 0.2.0 (2014-08-22)


#### Features

* **cli-options:** suppress email via --no-email option ([1e32ff42](https://github.com/vweevers/contributors-from-git/commit/1e32ff42cff94022a75acc4286e6d84db8d538a4))
* using _promises_ now

<a name="0.1.11"></a>
### 0.1.11 (2014-08-13)

* minor refactorings


<a name="0.1.10"></a>
### 0.1.10 (2014-06-09)


#### Bug Fixes

* **lib:** calculation of commits w/ different mail ([7ef93c01](https://github.com/vweevers/contributors-from-git/commit/7ef93c01c86a873b999b86583da37147c1dcc321), closes [#8](https://github.com/vweevers/contributors-from-git/issues/8))


<a name="0.1.9"></a>
### 0.1.9 (2014-06-05)


#### Features

* give at least 0.1 percent contribution ([8600dc08](https://github.com/vweevers/contributors-from-git/commit/8600dc084d87ed51e5e7328e5278852d5984856c), closes [#7](https://github.com/vweevers/contributors-from-git/issues/7))


<a name="0.1.6"></a>
### 0.1.6 (2014-05-19)


#### Features

* print usage example in case of invalid path ([064b52df](https://github.com/vweevers/contributors-from-git/commit/064b52df4740f9b3537449c850b663cf2178c34f))


<a name="0.1.5"></a>
### 0.1.5 (2014-04-21)


#### Bug Fixes

* **lib:** calculate commits based on email ([14d23f8e](https://github.com/vweevers/contributors-from-git/commit/14d23f8efba0502214e8150a1bf1f6789b1022e3), closes [#2](https://github.com/vweevers/contributors-from-git/issues/2))


<a name="0.1.4"></a>

### 0.1.4 (2014-04021)
* **change:** a _`/path/to/repository-dir/` is _mandatory_ now
* **test:** add fixture based testing

<a name="0.1.4"></a>

### 0.1.3 (2014-04-19)
* **lib:** minor refactorings


<a name="0.1.2"></a>

### 0.1.2 (2014-03-14)

#### Features

* **cli:** parse cli opts with "commander" ([ca382bd5](git://github.com/vweevers/contributors-from-git/commit/ca382bd5533e6eabb0ce1b5f11bb533c1b6d2840))


<a name="0.1.1"></a>

### 0.1.1 ()
* status badges


test/index.js
/* global describe, it */
'use strict'

// TODO: switch to tape
const chai = require('chai')
const expect = chai.expect
const tempy = require('tempy')
const fs = require('fs')
const exec = require('child_process').exec
const contributors = require('..')

// expect(x).to.be.true => expect(x).to.be.true()
chai.use(require('dirty-chai'))

function read (file) {
  return fs.readFileSync(file, 'utf8')
}

function mockLog (file) {
  return function (opts, callback) {
    fs.readFile(file, 'utf8', callback)
  }
}

describe('contributors-from-git', function () {
  describe('arguments', function () {
    it('should invoke callback', function (done) {
      contributors('.', function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.be.an('array')
        expect(result[0]).to.contain.keys(['commits', 'name', 'email'])
        done()
      })
    })

    it('throws when dir is not a string', function (done) {
      expect(contributors.bind(null, null, function () {})).to.throw()
      done()
    })

    it('yields an error if dir does not exist', function (done) {
      contributors('does-not-exist', function (err) {
        expect(err.code).to.equal('ENOENT')
        done()
      })
    })

    it('yields an error if dir is not a git working tree', function (done) {
      contributors(tempy.directory(), function (err) {
        expect(err.message).to.match(/Not a git repository/)
        done()
      })
    })

    it('yields an error if dir has no commits yet', function (done) {
      const dir = tempy.directory()

      exec('git init', { cwd: dir }, function (err) {
        if (err) throw err

        contributors(dir, function (err) {
          expect(err.message).to.match(/does not have any commits yet/)
          done()
        })
      })
    })
  })

  describe('contributors()', function () {
    it('can parse a single user commit', function (done) {
      const inFixture = 'test/fixtures/actual/single-user-single-commit.log'
      const outFixture = 'test/fixtures/expected/single-user-single-commit.json'

      contributors('.', { log: mockLog(inFixture) }, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(read(outFixture)))
        done()
      })
    })

    it('can parse multiple commits from a single user', function (done) {
      const inFixture = 'test/fixtures/actual/single-user-multiple-commit.log'
      const outFixture = 'test/fixtures/expected/single-user-multiple-commit.json'

      contributors('.', { log: mockLog(inFixture) }, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(read(outFixture)))
        done()
      })
    })

    it('can parse same user with different emails ', function (done) {
      const inFixture = 'test/fixtures/actual/single-user-multiple-commit-different-mail.log'
      const outFixture = 'test/fixtures/expected/single-user-multiple-commit-different-mail.json'

      contributors('.', { log: mockLog(inFixture) }, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(read(outFixture)))
        done()
      })
    })

    it('can parse multiple user with same email', function (done) {
      const inFixture = 'test/fixtures/actual/multi-user-same-mail.log'
      const outFixture = 'test/fixtures/expected/multi-user-same-mail.json'

      contributors('.', { log: mockLog(inFixture) }, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(read(outFixture)))
        done()
      })
    })
  })
})


test/test.markdown.js
/* global describe, it, beforeEach, afterEach */

'use strict'

var fs = require('fs')
var sinon = require('sinon')
var chai = require('chai')
var git = require('../lib/gitlog')
var expect = chai.expect
var GitContributors = require('..').GitContributors

// expect(x).to.be.true => expect(x).to.be.true()
chai.use(require('dirty-chai'))

var readIn = function readIn (file) {
  return fs.readFileSync(file, 'utf-8')
}

var stubFixture = function stubFixture (file) {
  sinon.stub(git, 'log').returns(readIn(file))
}

describe('git-sorted-contributors', function () {
  describe('with --markdown option', function () {
    var opts = { cwd: '.', markdown: true }

    beforeEach(function () {
    })

    afterEach(function () {
      git.log.restore()
    })

    it('can parse a single user commit', function (done) {
      var inFixture, outFixture

      inFixture = 'test/fixtures/actual/single-user-single-commit.log'
      outFixture = 'test/fixtures/expected/single-user-single-commit.md'

      stubFixture(inFixture, opts)

      GitContributors.list(opts, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.equal(readIn(outFixture))
        done()
      })
    })

    it('can parse multiple commits from a single user', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-multiple-commit.log'
      var outFixture = 'test/fixtures/expected/single-user-multiple-commit.md'

      stubFixture(inFixture, opts)

      GitContributors.list(opts, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.equal(readIn(outFixture))
        done()
      })
    })

    it('can parse same user with different emails', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-multiple-commit-different-mail.log'
      var outFixture = 'test/fixtures/expected/single-user-multiple-commit-different-mail.md'

      stubFixture(inFixture)

      GitContributors.list(opts, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.equal(readIn(outFixture))
        done()
      })
    })

    it('can parse options', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-multiple-commit-different-mail.log'
      var outFixture = 'test/fixtures/expected/single-user-multiple-commit-different-mail.md'

      stubFixture(inFixture, opts)

      GitContributors.list(opts, function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.equal(readIn(outFixture))
        done()
      })
    })
  })
})


test/test.parse-results.js
/* global describe, it, afterEach */
'use strict'

var fs = require('fs')
var sinon = require('sinon')
var chai = require('chai')
var git = require('../lib/gitlog')
var expect = chai.expect
var GitContributors = require('..').GitContributors

// expect(x).to.be.true => expect(x).to.be.true()
chai.use(require('dirty-chai'))

var readIn = function readIn (file) {
  return fs.readFileSync(file, 'utf-8')
}

var stubFixture = function stubFixture (file) {
  sinon.stub(git, 'log').returns(readIn(file))
}

describe('git-sorted-contributors', function () {
  describe('#list()', function () {
    afterEach(function () {
      git.log.restore()
    })

    it('can parse a single user commit', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-single-commit.log'
      var outFixture = 'test/fixtures/expected/single-user-single-commit.json'

      stubFixture(inFixture)

      GitContributors.list('.', function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(readIn(outFixture)))
        done()
      })
    })

    it('can parse multiple commits from a single user', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-multiple-commit.log'
      var outFixture = 'test/fixtures/expected/single-user-multiple-commit.json'

      stubFixture(inFixture)

      GitContributors.list('.', function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(readIn(outFixture)))
        done()
      })
    })

    it('can parse same user with different emails ', function (done) {
      var inFixture = 'test/fixtures/actual/single-user-multiple-commit-different-mail.log'
      var outFixture = 'test/fixtures/expected/single-user-multiple-commit-different-mail.json'

      stubFixture(inFixture)

      GitContributors.list('.', function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(readIn(outFixture)))
        done()
      })
    })

    it('can parse multiple user with same email', function (done) {
      var inFixture = 'test/fixtures/actual/multi-user-same-mail.log'
      var outFixture = 'test/fixtures/expected/multi-user-same-mail.json'

      stubFixture(inFixture)

      GitContributors.list('.', function (err, result) {
        expect(err).to.not.exist()
        expect(result).to.deep.equal(JSON.parse(readIn(outFixture)))
        done()
      })
    })
  })

  describe('when given wrong arguments', function () {
    it('should not throw when no path given via string', function (done) {
      var f = function () {
        GitContributors.list(null, function (/* err, result */) {
          done()
        })
      }

      expect(f).to.not.throw()
    })
  })
})


index.js
'use strict'

const path = require('path')
const exec = require('child_process').exec

function contributors (dir, opts, callback) {
  if (typeof opts === 'function') {
    callback = opts
    opts = null
  }

  const cwd = path.resolve(dir)
  const log = (opts && opts.log) || gitlog

  log({ cwd, maxBuffer: 32 * 1024 * 1024 }, function (err, stdout) {
    if (err) return callback(err)

    callback(null, parse(stdout))
  })
}

function gitlog (opts, callback) {
  exec('git log --pretty="%an %ae"', opts, callback)
}

function parse (stdout) {
  const entries = stdout.split('\n').filter(Boolean)
  const unique = Array.from(new Set(entries))

  return unique.map(function (committer) {
    const a = committer.split(' ')
    const email = a.pop()
    const name = a.join(' ')
    const commits = entries.filter(function (e) {
      return e.indexOf(name) > -1 && e.indexOf(email) > -1
    }).length

    return { name, email, commits }
  })
}

module.exports = contributors


lib/git-contributors.js
/*!
 * git-contributors (0.2.3) - 2015-06-14
 * https://github.com/vweevers/git-sorted-contributors.git
 * Copyright (c) 2014 David Linse; License: MIT
 */

/**
 * git-contributors module
 */

var fs    = require('fs'),
    path  = require('path'),
    git   = require('../lib/gitlog'),
    _     = require('lodash');
var Q     = require('q');

var verifyRepositoryExists = function() {
  'use strict';

  var deferred = Q.defer();
  var repo     = path.join(program.cwd, '.git');
  var ok       = fs.existsSync(program.cwd) && fs.existsSync(repo);

  if (ok) { deferred.resolve(program); }
  else {
    deferred.reject(
      new Error('Could not find .git repository at "'+ program.cwd +'"')
    );
  }

  return deferred.promise;
};

var processLog = function(stdout) {
  'use strict';

  var deferred = Q.defer();

  var list,
      entries = _.compact(stdout.split('\n')),
      total   = _.size(entries);

  list = _.map(_.uniq(entries), function(committer) {

    var parts   = committer.split(' '),
        email   = _.last(parts),
        author  = _.without(parts, email).join(' '),
        commits = _.size(_.filter(entries, function(e) {
          return e.indexOf(author) > -1 && e.indexOf(email) > -1;
        })),
        percentage = (commits / total * 100);

    return {
      commits: commits,
      name:    author,
      email:   email,
      percent: parseFloat(Math.max(0.1, percentage).toFixed(1), 10)
    };
  });

  list = _.sortBy(list, function(committer) {
    return -committer.commits;
  });

  deferred.resolve(list);

  return deferred.promise;
};

var format = function(data) {
  'use strict';

  var deferred = Q.defer();

  if (program.markdown === true) {
    data = require('../lib/markdown-reporter').format(data, program);
  }

  if (program.json === true) {
    data = JSON.stringify(data);
  }
  deferred.resolve(data);

  return deferred.promise;
};

var filter = function(data) {
  'use strict';

  var deferred = Q.defer();

  var stripEmail = function(el) {
    return _.omit(el, 'email');
  };

  if (program.email === false) {
    data = _.map(data, stripEmail);
  }

  deferred.resolve(data);

  return deferred.promise;
};

//--

var program = {
  timeout: 5000,
  cwd: '.',
  maxBuffer: 25000 * 1024
};


var GitContributors = function GitContributors () {};

GitContributors.prototype.list = function(opts, cb) {
  'use strict';

  program.markdown = null;
  program.json = null;

  if (_.isString(opts)) {
    program.cwd = opts;
  }
  else {
    program = _.merge(program, opts);
  }

  Q()
  .then(verifyRepositoryExists)
  .then(git.log)
  .then(processLog)
  .then(filter)
  .then(format)
  .done(function(result) {
    cb(null, result);
  }, function(err) {
    cb(err, null);
  });
};

module.exports.GitContributors = new GitContributors();
